Packages/Core/Editor/Drawers/AtomBaseReferenceDrawer.cs

```diff
@@ -91,14 +119,98 @@ public override void OnGUI(Rect position, SerializedProperty property, GUIConten
                 }
                 else
                 {
-                    EditorGUI.PropertyField(position, usageTypeProperty, GUIContent.none);
+                    EditorGUI.PropertyField(guiData.Position, usageTypeProperty, GUIContent.none);
                 }
             }
-            if (EditorGUI.EndChangeCheck())
-                property.serializedObject.ApplyModifiedProperties();
+        }
+        
+        private static void SetUsageIndex(SerializedProperty property, int index)
+        {
+            property.FindPropertyRelative(USAGE_PROPERTY_NAME).intValue = index;
+        }
+        
+        private static int GetUsageIndex(SerializedProperty property)
+        {
+            return property.FindPropertyRelative(USAGE_PROPERTY_NAME).intValue;
+        }
+        
+        
+        #region Auto Drag And Drop Usage Type Detection
+        private void DetermineDragAndDropFieldReferenceType(in GuiData guiData)
+        {
+            EventType mouseEventType = Event.current.type;
+
+            if (mouseEventType != EventType.DragUpdated && mouseEventType != EventType.DragPerform)
+                return;
+            
+            if (!IsMouseHoveringOverProperty(guiData.Position))
+                return;
+
+            var draggedObjects = DragAndDrop.objectReferences;
+
+            if (draggedObjects.Length < 1)
+                return;
+            
+            Object draggedObject = draggedObjects[0];
+
+            if (draggedObject is GameObject gameObject)
+            {
+                object[] instancers = gameObject.GetComponents<IAtomInstancer>();
+
+                UpdateUsageConfigurationOption(guiData.Property, instancers);
+            }
+            else
+                UpdateUsageConfigurationOption(guiData.Property, draggedObject);
+        }
+        
+        private void UpdateUsageConfigurationOption(SerializedProperty property, params object[] draggedObjects)
+        {
+            if (draggedObjects == null || draggedObjects.Length < 1)
+                return;
+            
+            var usages = GetUsages(property);
+            int currentUsageIndex = GetUsageIndex(property);
+            int newUsageIndex = -1;
+            
+            foreach (object draggedObject in draggedObjects)
+            {
+                string draggedObjectType = draggedObject.GetType().Name;
 
-            EditorGUI.indentLevel = indent;
-            EditorGUI.EndProperty();
+                for (int index = 0; index < usages.Length; index++)
+                {
+                    var usage = usages[index];
+                    SerializedProperty fieldProperty = property.FindPropertyRelative(usage.PropertyName);
+                    string fieldPropertyType = fieldProperty.type.Replace("PPtr<$", "").Replace(">", "");
```

Comment:
This is quite hard to understand without context. Can we add a comment on why we do this here?

Reply:
I pushed a change where I refactored this section of code to hopefully make it easier to understand without adding comments

Reply:
The reason why I compare strings and not types directly is because I couldn't figure out how to get the type of the `SerializedProperty` other than using it's `type` which is a string. It adds additional `PPtr<$...>` wrapper around the type name but other than that it does the job. If you have an alternative way that works I don't mind changing that code once again.

Reply:
Great, thanks! I believe the reasoning behind it is more clear now. 

FYI. I took the liberty to do some tweaks to the PR and took into account if the type used by the generic is in itself generic (even though that might never be the case atm, but never the less feels good to secure for that just to be sure).